File: techniques/polychoric-correlation/r/polychoric_correlation.R
# Tetrachoric and polychoric correlation (Reference §4.7)
# From-scratch base R plus polycor::polychor / psych::polychoric.
# Run with:  Rscript polychoric_correlation.R
#
# Inputs used below:
#   table : r x c matrix of nonnegative integer counts (ordinal-by-ordinal)

thresholds_from_marginals <- function(margins) {
  cum <- cumsum(margins); cum <- cum[-length(cum)]
  qnorm(pmin(pmax(cum, 1e-8), 1 - 1e-8))
}

cell_prob <- function(a_lo, a_hi, b_lo, b_hi, rho) {
  bvn <- function(a, b) mvtnorm::pmvnorm(upper = c(a, b),
                                          corr = matrix(c(1, rho, rho, 1), 2))
  if (!requireNamespace("mvtnorm", quietly = TRUE))
    stop("install 'mvtnorm' for bivariate normal CDF")
  a_hi <- if (is.infinite(a_hi)) 8 else a_hi
  a_lo <- if (is.infinite(a_lo)) -8 else a_lo
  b_hi <- if (is.infinite(b_hi)) 8 else b_hi
  b_lo <- if (is.infinite(b_lo)) -8 else b_lo
  as.numeric(bvn(a_hi, b_hi) - bvn(a_lo, b_hi) - bvn(a_hi, b_lo) + bvn(a_lo, b_lo))
}

polychoric_scratch <- function(table) {
  obs <- as.matrix(table); r <- nrow(obs); c <- ncol(obs); n <- sum(obs)
  row_m <- rowSums(obs) / n; col_m <- colSums(obs) / n
  alpha <- c(-Inf, thresholds_from_marginals(row_m), Inf)
  beta  <- c(-Inf, thresholds_from_marginals(col_m), Inf)
  neg_ll <- function(rho) {
    ll <- 0
    for (i in seq_len(r)) for (j in seq_len(c)) {
      if (obs[i, j] == 0) next
      p <- cell_prob(alpha[i], alpha[i + 1], beta[j], beta[j + 1], rho)
      ll <- ll + obs[i, j] * log(max(p, 1e-15))
    }
    -ll
  }
  res <- optimize(neg_ll, c(-0.999, 0.999), tol = 1e-6)
  list(rho = res$minimum, row_thresholds = alpha[2:(r)], col_thresholds = beta[2:(c)],
       log_likelihood = -res$objective, n = n, size = c(r, c),
       method = if (r == 2 && c == 2) "tetrachoric" else "polychoric")
}

# Library: polycor::polychor (or polycor::polychor for 2x2 = tetrachoric),
#          psych::polychoric (also returns SEs and a correlation matrix)
library_demo <- function(table) {
  if (requireNamespace("polycor", quietly = TRUE)) {
    cat("polycor::polychor:\n"); print(polycor::polychor(table, std.err = TRUE))
  } else cat("(install 'polycor' for polychor)\n")
}

if (sys.nframe() == 0) {
  table <- matrix(c(20, 8, 2, 12, 30, 10, 3, 12, 25), nrow = 3, byrow = TRUE)
  cat("=== Polychoric (3x3) ===\n");  print(polychoric_scratch(table))
  cat("\n=== Tetrachoric (2x2) ===\n")
  print(polychoric_scratch(matrix(c(60, 15, 10, 40), nrow = 2, byrow = TRUE)))
  cat("\n--- library ---\n"); library_demo(table)
}

File: techniques/polychoric-correlation/python/polychoric_correlation.py
"""Tetrachoric and polychoric correlation (Reference §4.7).

Model: an unobserved bivariate normal (X*, Y*) generates the observed
ordinal pair (X, Y) by thresholds:
    X = i  iff  alpha_{i-1} < X* <= alpha_i,    i = 1..r
    Y = j  iff  beta_{j-1}  < Y* <= beta_j,     j = 1..c
with alpha_0 = beta_0 = -inf, alpha_r = beta_c = +inf.

The *polychoric correlation* is the correlation rho of (X*, Y*) -- the
"underlying" association on the latent continuous scale. The thresholds are
estimated from the marginals (via Phi^{-1}(cumulative marginal proportions)),
and rho is found by maximizing the log-likelihood of the multinomial cell
counts.

Special case: 2x2 tables = *tetrachoric* correlation.

This implementation
-------------------
- Estimates thresholds from observed marginals: alpha_i = Phi^{-1}( sum_{j<=i} p_i. )
- Maximizes the bivariate-normal cell-probability log-likelihood over rho via
  scipy.optimize on [-0.999, 0.999].
- Returns rho-hat plus the threshold vectors. Standard errors require numerical
  derivatives; we do not estimate them here -- use R's polycor::polychor or
  Python's semopy for SEs.
"""
from __future__ import annotations

import math
from typing import Sequence

import numpy as np
from scipy import optimize, stats


def _thresholds_from_marginals(margins: Sequence[float]) -> np.ndarray:
    """alpha_i = Phi^-1(cum proportion up to i)."""
    cum = np.cumsum(np.asarray(margins, dtype=float))
    cum = cum[:-1]                                # drop the trailing 1.0
    cum = np.clip(cum, 1e-8, 1 - 1e-8)
    return stats.norm.ppf(cum)


def _bvn_cdf_orthant(a, b, rho):
    """P(X* <= a, Y* <= b) for a standard bivariate normal with correlation rho."""
    return stats.multivariate_normal.cdf([a, b], mean=[0, 0],
                                         cov=[[1, rho], [rho, 1]])


def _cell_prob(alpha_lo, alpha_hi, beta_lo, beta_hi, rho):
    """P(alpha_lo < X* <= alpha_hi, beta_lo < Y* <= beta_hi) via inclusion/exclusion."""
    a_hi = alpha_hi if math.isfinite(alpha_hi) else 8.0
    a_lo = alpha_lo if math.isfinite(alpha_lo) else -8.0
    b_hi = beta_hi if math.isfinite(beta_hi) else 8.0
    b_lo = beta_lo if math.isfinite(beta_lo) else -8.0
    return (_bvn_cdf_orthant(a_hi, b_hi, rho)
            - _bvn_cdf_orthant(a_lo, b_hi, rho)
            - _bvn_cdf_orthant(a_hi, b_lo, rho)
            + _bvn_cdf_orthant(a_lo, b_lo, rho))


def polychoric(table) -> dict:
    """Maximum-likelihood polychoric correlation for an r x c table of counts.

    ``table`` is a 2D array of nonnegative integer counts.
    """
    obs = np.asarray(table, dtype=float)
    r, c = obs.shape
    n = obs.sum()
    row_marg = obs.sum(axis=1) / n
    col_marg = obs.sum(axis=0) / n
    alpha = np.concatenate([[-np.inf], _thresholds_from_marginals(row_marg), [np.inf]])
    beta = np.concatenate([[-np.inf], _thresholds_from_marginals(col_marg), [np.inf]])

    def neg_loglik(rho):
        ll = 0.0
        for i in range(r):
            for j in range(c):
                if obs[i, j] == 0:
                    continue
                p = _cell_prob(alpha[i], alpha[i + 1], beta[j], beta[j + 1], rho)
                p = max(p, 1e-15)
                ll += obs[i, j] * math.log(p)
        return -ll

    res = optimize.minimize_scalar(neg_loglik, bounds=(-0.999, 0.999), method="bounded")
    return {"rho": float(res.x),
            "row_thresholds": alpha[1:-1].tolist(),
            "col_thresholds": beta[1:-1].tolist(),
            "log_likelihood": float(-res.fun),
            "n": float(n), "size": (r, c),
            "method": "tetrachoric" if (r, c) == (2, 2) else "polychoric"}


def library_versions(table):
    out = {}
    try:
        from semopy.polycorr import polychoric_corr  # type: ignore
        out["semopy polychoric_corr"] = float(polychoric_corr(np.asarray(table)))
    except Exception as exc:
        out["semopy"] = f"not used ({exc})"
    return out


if __name__ == "__main__":
    # Example: 3 x 3 ordinal table (Likert vs. Likert)
    table = np.array([
        [20,  8,  2],
        [12, 30, 10],
        [ 3, 12, 25],
    ])
    print("=== Polychoric on 3x3 ordinal table ===")
    res = polychoric(table)
    for k, v in res.items():
        print(f"  {k:18s}: {v}")

    # Tetrachoric (2x2 special case)
    table2 = np.array([[60, 15], [10, 40]])
    print("\n=== Tetrachoric on 2x2 ===")
    for k, v in polychoric(table2).items():
        print(f"  {k:18s}: {v}")

    print("\n--- library ---")
    for k, v in library_versions(table).items():
        print(f"  {k}: {v}")
